orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -787,7 +788,9 @@ def _training_session_run_helper(self, is_train, inputs, inputs_desc, outputs_de
         outputs_desc_resolved = self._resolve_symbolic_dimensions(inputs, inputs_desc, outputs_desc)
         result = {}
         for output_desc in outputs_desc_resolved:
-            torch_tensor = torch.zeros(output_desc.shape, device=self.options.device.id,
+            target_device = self._select_device_if_general_or_cpu_if_allfinite(
+                self.options.device.id, output_desc.name)
+            torch_tensor = torch.zeros(output_desc.shape, device=target_device,
```

Comment:
An alternative to create a 3 line funtions that is used once is to default to the user-specified device and change it on a single `if`
```python
            target_device = self.options.device.id
            if self.options.mixed_precision.enabled and name == self.model_desc.all_finite.name:
                # Keep all finite flag on CPU to match backend implementation
                # This prevents CPU -> GPU -> CPU copies between frontend and backend
                target_device = 'cpu'
            torch_tensor = torch.zeros(output_desc.shape, device=target_device,`
```

Reply:
Modified as above in PR at https://github.com/microsoft/onnxruntime/pull/5371